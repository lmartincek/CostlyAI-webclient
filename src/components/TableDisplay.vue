<script setup lang="ts">
// const props = defineProps(['data'])
defineProps(['data'])
</script>

<template>
    <div>
<!--        <h2>{{category}}</h2>-->

        <table v-if="data" border="1">
            <thead>
            <tr>
                <th>Item</th>
                <th>Price</th>
            </tr>
            </thead>
            <tbody>

            <tr v-for="(value, key) in data" :key="key">
                <td>{{ value.name }}</td>
                <!-- TODO - currency based on country, add currency file-->
                <td>{{ Number(value.price).toFixed(2) }} $</td>
            </tr>
            </tbody>
        </table>

        <p v-else>No data available</p>
    </div>
</template>
