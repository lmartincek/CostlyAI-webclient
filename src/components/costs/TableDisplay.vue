<script setup lang="ts">
defineProps(['data', 'category'])
</script>

<template>
  <div class="table-wrapper">
    <div class="table-wrapper__header">
      {{ category }}
    </div>
    <div class="table-wrapper__body">
      <div class="row">
        <span><b>Item</b></span>
        <span><b>Price</b></span>
      </div>
      <div v-for="(value, key) in data" :key="key" class="row">
        <span class="name">{{ value.name }}</span>
        <span class="price"
          ><b>${{ value.price.toFixed(2) }} USD</b></span
        >
      </div>
    </div>
  </div>
</template>

<style lang="scss">
.table-wrapper {
  color: #000;
  box-shadow: 0 4px 8px rgba(0, 0, 0, 0.1);
  background-color: #fff;
  margin: 1rem 0 1.5rem;
  border-radius: 1rem;
  flex: 1 1 300px;
  min-width: 200px;

  &__header {
    background: $primary-gradient;
    color: $white-color;
    font-size: 1.15rem;
    font-weight: 500;
    padding: 1rem 1.5rem;
    text-align: left;
    border-radius: 1rem 1rem 0 0;

    &::first-letter {
      text-transform: capitalize;
    }
  }

  &__body {
    color: $text-color;

    .row {
      margin: 1rem 1.5rem 0;
      padding: 0 0 1rem;
      display: flex;
      justify-content: space-between;
      align-items: center;
      border-bottom: 1px solid $border-color;

      &:last-of-type {
        border-bottom: none;
        margin-bottom: 0.5rem;
      }

      &:first-of-type {
        margin-top: 1.5rem;
      }

      .name {
        text-align: left;
      }

      .price {
        text-align: right;
      }
    }
  }
}
</style>
