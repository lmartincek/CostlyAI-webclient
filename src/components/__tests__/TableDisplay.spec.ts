import { describe, it, expect } from 'vitest'
import { mount } from '@vue/test-utils'
import TableDisplay from '@/components/costs/TableDisplay.vue'

describe('DataTable Component', () => {
  const mockData = {
    item1: { name: 'Product A', price: 19.99 },
    item2: { name: 'Product B', price: 29.99 },
    item3: { name: 'Product C', price: 39.99 }
  }

  const mockCategory = 'Electronics'

  it('renders the table with correct structure', () => {
    const wrapper = mount(TableDisplay, {
      props: {
        data: mockData,
        category: mockCategory
      }
    })

    expect(wrapper.find('table.table-wrapper').exists()).toBe(true)
    expect(wrapper.find('thead.table-wrapper__header').exists()).toBe(true)
    expect(wrapper.find('tbody.table-wrapper__body').exists()).toBe(true)
  })

  it('displays the correct category in header', () => {
    const wrapper = mount(TableDisplay, {
      props: {
        data: mockData,
        category: mockCategory
      }
    })

    const header = wrapper.find('th')
    expect(header.text()).toBe(mockCategory)
    expect(header.attributes('aria-label')).toBe('Product category')
  })

  it('renders column headers correctly', () => {
    const wrapper = mount(TableDisplay, {
      props: {
        data: mockData,
        category: mockCategory
      }
    })

    const headers = wrapper.findAll('tbody tr.row').at(0)?.findAll('td')
    expect(headers?.length).toBe(2)
    expect(headers?.at(0)?.text()).toBe('Item')
    expect(headers?.at(1)?.text()).toBe('Price')
  })

  it('renders all data rows correctly', () => {
    const wrapper = mount(TableDisplay, {
      props: {
        data: mockData,
        category: mockCategory
      }
    })

    const rows = wrapper.findAll('tbody tr.row').slice(1)
    expect(rows.length).toBe(Object.keys(mockData).length)

    rows.forEach((row, index) => {
      const cells = row.findAll('td')
      const dataKey = Object.keys(mockData)[index]
      const item = mockData[dataKey as keyof typeof mockData]

      expect(cells[0].classes()).toContain('name')
      expect(cells[0].text()).toBe(item.name)
      expect(cells[0].attributes('aria-label')).toBe('Product name')

      expect(cells[1].classes()).toContain('price')
      expect(cells[1].text()).toBe(`$${item.price.toFixed(2)} USD`)
      expect(cells[1].attributes('aria-label')).toBe('Product price')
    })
  })

  it('handles empty data prop gracefully', () => {
    const wrapper = mount(TableDisplay, {
      props: {
        data: {},
        category: mockCategory
      }
    })

    const rows = wrapper.findAll('tbody tr.row')
    expect(rows.length).toBe(1)
  })

  it('formats prices correctly with 2 decimal places', () => {
    const testData = {
      item1: { name: 'Test Product', price: 20 }
    }

    const wrapper = mount(TableDisplay, {
      props: {
        data: testData,
        category: mockCategory
      }
    })

    const priceCell = wrapper.find('td.price')
    expect(priceCell.text()).toBe('$20.00 USD')
  })

  it('matches snapshot with data', () => {
    const wrapper = mount(TableDisplay, {
      props: {
        data: mockData,
        category: mockCategory
      }
    })

    expect(wrapper.html()).toMatchSnapshot()
  })

  it('matches snapshot with empty data', () => {
    const wrapper = mount(TableDisplay, {
      props: {
        data: {},
        category: mockCategory
      }
    })

    expect(wrapper.html()).toMatchSnapshot()
  })
})
