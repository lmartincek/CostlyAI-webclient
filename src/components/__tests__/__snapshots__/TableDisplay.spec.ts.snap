// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`DataTable Component > matches snapshot with data 1`] = `
"<table class="table-wrapper">
  <thead class="table-wrapper__header">
    <tr>
      <th aria-label="Product category">Electronics</th>
    </tr>
  </thead>
  <tbody class="table-wrapper__body">
    <tr class="row">
      <td><b>Item</b></td>
      <td><b>Price</b></td>
    </tr>
    <tr class="row">
      <td class="name" aria-label="Product name">Product A</td>
      <td class="price" aria-label="Product price"><b>$19.99 USD</b></td>
    </tr>
    <tr class="row">
      <td class="name" aria-label="Product name">Product B</td>
      <td class="price" aria-label="Product price"><b>$29.99 USD</b></td>
    </tr>
    <tr class="row">
      <td class="name" aria-label="Product name">Product C</td>
      <td class="price" aria-label="Product price"><b>$39.99 USD</b></td>
    </tr>
  </tbody>
</table>"
`;

exports[`DataTable Component > matches snapshot with empty data 1`] = `
"<table class="table-wrapper">
  <thead class="table-wrapper__header">
    <tr>
      <th aria-label="Product category">Electronics</th>
    </tr>
  </thead>
  <tbody class="table-wrapper__body">
    <tr class="row">
      <td><b>Item</b></td>
      <td><b>Price</b></td>
    </tr>
  </tbody>
</table>"
`;
